internal/broker/broker.go

```diff
@@ -475,12 +486,51 @@ func (b *Broker) handleIsAuthenticated(ctx context.Context, session *sessionInfo
 		return AuthNext, nil
 
 	case "password":
-		authInfo, err = b.loadAuthInfo(ctx, session, challenge)
+		var useOldEncryptedToken bool
+		exists, err := utils.FileExists(session.passwordPath)
+		if err != nil {
+			slog.Error(err.Error())
+			return AuthDenied, errorMessage{Message: "could not check password file"}
+		}
+		if !exists {
+			// For backwards compatibility, we also check the old encrypted token path.
+			exists, err = utils.FileExists(session.oldEncryptedTokenPath)
+			if err != nil {
+				slog.Error(err.Error())
+				return AuthDenied, errorMessage{Message: "could not check old encrypted token path"}
+			}
+			if !exists {
+				return AuthDenied, errorMessage{Message: "password file does not exist"}
+			}
+			useOldEncryptedToken = true
+		}
+
+		if !useOldEncryptedToken {
+			ok, err := b.checkPassword(challenge, session.passwordPath)
+			if err != nil {
+				slog.Error(err.Error())
+				return AuthRetry, errorMessage{Message: "could not check password"}
+			}
+			if !ok {
+				return AuthDenied, errorMessage{Message: "invalid password"}
+			}
+		}
+
+		authInfo, err = b.loadAuthInfo(ctx, session, challenge, useOldEncryptedToken)
 		if err != nil {
 			slog.Error(err.Error())
 			return AuthRetry, errorMessage{Message: "could not load cached info"}
 		}
 
+		if useOldEncryptedToken {
+			// We were able to decrypt the old token with the password, so we can now hash and store the password in the
+			// new format.
+			if err = b.hashAndStorePassword(challenge, session.passwordPath); err != nil {
+				slog.Error(err.Error())
+				return AuthDenied, errorMessage{Message: "could not store password"}
```

Comment:
> As a rule of thumb, we don’t log the error and return it at each point.

And we also don't do that here. We log the original error and return a generic error message.

> I suggest that we use a similar mechanism here to the redactError in authd: We wrap the initial error with the error we want to return outside to not leak information.
> 
> Then, at the edge of the return caller (in a defer function), we check for the error type and then log the complete error + return the redacted one.
> 
> That way, you ensure that you don’t forget one spot you didn’t log the error before redacting (I’m seeing a lot of slog.Error() + "returning an error" in that function for instance.

I see some problems with that approach, but I'm not sure this PR is the best place to discuss that, because it doesn't actually do anything different regarding the error handling in the broker, it simply follows what we're already doing. Should we move that discussion somewhere else?

Reply:
> And we also don't do that here. We log the original error and return a generic error message.

I agree, but we still handle one error which mirrored to a redacted error.

> I see some problems with that approach, but I'm not sure this PR is the best place to discuss that, because it doesn't actually do anything different regarding the error handling in the broker, it simply follows what we're already doing. Should we move that discussion somewhere else?

Definitively, I’m happy with that part as is, and then, I’ll do another PR with what I had in mind (and change what was merged too during my holidays about the error logging).